File: apps/retro-platform-web/src/domain/room.ts
export type RoomStatus = 'LOBBY' | 'GAME_SELECTION' | 'PLAYING' | 'FINISHED' | 'CLOSED';

export interface RoomPlayer {
  id: string;
  displayName: string;
  color: string;
  /** A picked pixel-art portrait id (see utils/avatarOptions.js); absent means the initials avatar is shown instead. */
  avatarId?: string;
  isHost: boolean;
  isReady: boolean;
  isConnected?: boolean;
  joinedAt?: number;
}

/** Records that a draw was settled at random, so the UI can show it happening. */
export interface TieBreak {
  candidates: string[];
  winner: string;
}

export interface RetroRoom {
  id: string;
  code: string;
  roomName: string;
  hostPlayerId: string;
  players: RoomPlayer[];
  selectedGameId?: string;
  /** playerId -> gameId, collected during GAME_SELECTION. */
  votes?: Record<string, string>;
  /** Epoch ms when the vote closes; absent once the vote is resolved. */
  votingEndsAt?: number;
  /** Epoch ms when the room-wide vote opened. */
  votingStartedAt?: number;
  /** The games this round may choose between, so the server can close the vote itself. */
  candidateGameIds?: string[];
  tieBreak?: TieBreak;
  status: RoomStatus;
  maxParticipants: number;
  questionTimeSeconds: number;
  votingTimeSeconds: number;
  fileName?: string;
  description?: string;
  createdAt: number;
  currentGameSession?: {
    gameSessionId: string;
    gameId: string;
    roundId: string;
    seed: number;
    roundStartAtUnixMs: number | null;
    state: 'ACTIVE' | 'ENDED';
  };
}

export interface CreateRoomRequest {
  displayName: string;
  color: string;
  avatarId?: string;
  roomName: string;
  maxParticipants: number;
  questionTimeSeconds: number;
  votingTimeSeconds: number;
  fileName?: string;
  description?: string;
}

export interface JoinRoomRequest {
  roomCode: string;
  displayName: string;
  color: string;
  avatarId?: string;
}

export interface CreateRoomResult {
  room: RetroRoom;
  player: RoomPlayer;
  reconnectToken?: string;
}

export type JoinRoomErrorCode =
  | 'ROOM_NOT_FOUND'
  | 'ROOM_FULL'
  | 'ROOM_ALREADY_STARTED'
  | 'INVALID_ROOM_CODE';

export type JoinRoomResult =
  | { ok: true; room: RetroRoom; player: RoomPlayer; reconnectToken?: string }
  | { ok: false; error: JoinRoomErrorCode };

File: services/retro-platform-api/Domain/Room.cs
namespace RetroPlatform.Api.Domain;

/// <summary>
/// Room lifecycle values. These are plain strings rather than an enum because
/// they cross the wire into the TypeScript union in <c>domain/room.ts</c>;
/// keeping them literal removes any chance of a serialization mismatch.
/// </summary>
public static class RoomStatus
{
    public const string Lobby = "LOBBY";
    public const string GameSelection = "GAME_SELECTION";
    public const string Playing = "PLAYING";
    public const string Finished = "FINISHED";
}

public sealed record RoomPlayer
{
    public required string Id { get; init; }
    public required string DisplayName { get; init; }
    public required string Color { get; init; }
    public required bool IsHost { get; init; }
    public required bool IsReady { get; init; }

    /// <summary>
    /// Unix ms when this player's connection dropped, or null while connected.
    /// A refresh drops the socket too, so players are kept in the room for a
    /// short grace period instead of being removed the moment they disappear.
    /// </summary>
    public long? DisconnectedAt { get; init; }
}

public sealed record TieBreak
{
    public required IReadOnlyList<string> Candidates { get; init; }
    public required string Winner { get; init; }
}

/// <summary>
/// Mirrors <c>RetroRoom</c> in the web app. Immutable: every change produces a
/// new snapshot, which is what gets broadcast to the room's SignalR group.
/// </summary>
public sealed record RetroRoom
{
    public required string Id { get; init; }
    public required string Code { get; init; }
    public required string RoomName { get; init; }
    public required string HostPlayerId { get; init; }
    public required IReadOnlyList<RoomPlayer> Players { get; init; }
    public required string Status { get; init; }
    public required int MaxParticipants { get; init; }
    public required int QuestionTimeSeconds { get; init; }
    public required int VotingTimeSeconds { get; init; }
    public required long CreatedAt { get; init; }

    public string? SelectedGameId { get; init; }
    /// <summary>playerId -> gameId, collected during GAME_SELECTION.</summary>
    public IReadOnlyDictionary<string, string>? Votes { get; init; }
    /// <summary>Unix ms when the vote closes; null once it is resolved.</summary>
    public long? VotingEndsAt { get; init; }
    /// <summary>
    /// The games on offer this round, supplied when the vote opens so the
    /// server can close it on time without knowing the client's catalogue.
    /// TODO: move the catalogue itself server-side once games are registered there.
    /// </summary>
    public IReadOnlyList<string>? CandidateGameIds { get; init; }
    public TieBreak? TieBreak { get; init; }
    public string? FileName { get; init; }
    public string? Description { get; init; }
}
